### PR TITLE
fix(Video): stream the video transcode instead of buffering every frame in RAM (CORE-353) (CORE-351)

### PR DESCRIPTION
**Pull request by bigcat88**
`VideoFromFile.save_to()` has two paths: remux (stream copy) when source already matches the requested container/codec, and a transcode otherwise
The transcode first materialized the entire video as float32 tensors (`get_components` --> `VideoFromComponents.save_to`), with peak memory:

```
peak = 2 * frames * width * height * 12 bytes      (~ 47.5 MiB per 1080p frame)
```

of the file actual size. Every Partner Node that uploads video hits this path whenever the source is not already H.264

Note: The video cropping operation also forces the transcoding path (for any codec), so this PR also improves the situation when the `Video Slice` node is used

## Changes in this PR

Transcode branch is rewritten as a single streaming pass: demux --> decode --> reformat --> H.264/AAC encode --> mux, one frame at a time (`_save_transcoded`). Peak memory no longer scales with video length

Preserved behavior, now verified by tests:

* trim windows (`Trim Video`) - video and audio, sample-exact, including chained trims
* display-matrix rotation is baked in (portrait videos), dimensions swap for 90/270
* full-range sources (`yuvj`/MJPEG) are compressed to limited range as before
* 10-bit sources stay `yuv420p10le`
* audio is clamped to the video span; undecodable audio tracks (e.g. iPhone APAC spatial audio) are skipped with a warning as it originally was
* workflow metadata (prompt / extra_pnginfo) is written to the output as before
* the output opens lazily on the first kept frame, so cluster-muxed sources (webm/mkv) don't lose their leading audio block to a rewind

### Behavior changes (transcode path only)

* VFR sources keep their real frame timing (pts pass through, rebased to 0) instead of being re-timed to the average rate; audio is clamped to the true encoded span
* source metadata now survives a transcode (it always survived a remux)
* 7.1/8-channel audio downmixes to stereo - the old path crashed (segfault on 8-channel)
* clear, early errors instead of deep PyAV failures: no decodable frames, odd dimensions, mid-stream resolution changes; broken mid-stream timestamps are nudged monotonic instead of failing at mux
* 10-bit -> 8-bit reduction uses `swscale` dithering instead of float truncation
* the H.264 output is encoded without B-frames (`max_b_frames = 0`): with B-frame reordering the mp4 muxer derives sample durations from decode order, which broke irregular-VFR spans and trim windows; output is a few percent larger for it

<details>
<summary>Load Video --> Save Video benchmarks</summary>

Screenshot From 2026-07-07 18-46-01

Screenshot From 2026-07-07 18-46-42

The speedup has the same root as the memory win: ~75 % of the old wall time was spent converting YUV --> float RGB --> torch tensors --> RGB --> YUV
</details>

**Fixes CORE-353**
**Related to CORE-351** (the `Get Video Components` eager tensor materialization is a separate issue, not addressed here)

### CODE REVIEW

File: tests-unit/comfy_api_test/video_types_test.py [hunk 1]
```diff
@@ -237,3 +238,386 @@ def test_duration_consistency(video_components):
...
+            for offset in range(0, 44100 * frames // fps, 1024):
+                n = min(1024, 44100 * frames // fps - offset)
+                audio_frame = av.AudioFrame.from_ndarray(
+                    torch.zeros(1, n, dtype=torch.int16).numpy(), format="s16", layout="mono"
+                )
+                audio_frame.sample_rate = 44100
+                audio_frame.pts = offset
+                container.mux(stream.encode(audio_frame))
+        for stream in [video_stream, *audio]:
+            container.mux(stream.encode(None))
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply tests-unit/comfy_api_test/video_types_test.py:283]: I don't think this one reproduces on the supported PyAV range

This path is already covered by `test_save_to_transcode_trims_audio_in_stream_time_base_units()`, which calls:

```
create_transcode_source(frames=90, container_format="matroska", audio_codec="aac")
```

So if `create_transcode_source()` could not encode the `s16` test audio into AAC, the test would fail during source setup before reaching the actual transcode code

I also checked this manually across PyAV versions `16.0.0`, `17.0.1`, and `18.0.0`, and the helper produces a valid decodable AAC stream in all of them

Can you confirm whether you have a concrete failing repro for this comment, or was this based on the general expectation that AAC encoders require `fltp` input?

File: comfy_api/latest/_input_impl/video_types.py [hunk 2]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                                tb = resampled.time_base if resampled.time_base else audio_time_base
+                                frame_start = float(resampled.pts * tb)
+                            to_skip = max(0, int((start_time - frame_start) * sample_rate))
+                            if to_skip >= resampled.samples:
+                                continue
+                            audio_started = True
+                            if to_skip:
+                                pending_audio.append(audio_frame_from_ndarray(resampled.to_ndarray()[..., to_skip:]))
+                                continue
+                        pending_audio.append(resampled)
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:751]: This is a fair caveat for sparse media, but I don't think it should block this PR, as the PR makes the situation much better than it was

The previous path always materialized the full decoded video/audio before encoding. This PR fixes that practical OOM path and streams normal dense transcodes. `pending_audio` can still grow for pathological cases where audio demuxes far ahead of video progress, such as very sparse video frames over long continuous audio, but that is n rare edge case and fixing it cleanly would require a more complex approach

File: comfy_api/latest/_input_impl/video_types.py [hunk 3]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                if packet.stream == video_stream and not video_done:
+                    try:
+                        frames = packet.decode()
+                    except av.error.InvalidDataError:
+                        logging.info("pyav decode error")
+                        continue
+                    for frame in frames:
+                        if frame.pts is not None and frame.pts < start_pts:
+                            continue
+                        if end_pts is not None and frame.pts is not None and frame.pts >= end_pts:
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:631]: I checked the same missing-PTS scenario against `origin/master`, and the old transcode path did not handle it correctly either - it raises `TypeError` on the first pts-less frame at the trim comparison

With this PR, normal timestamped videos trim correctly, and the mixed missing/real PTS crash case is now guarded. Fully pts-less video would still only be approximate because there is no source timeline to trim against

Ignore such malformed-media for now - it will add much complexity to handle this

File: comfy_api/latest/_input_impl/video_types.py [hunk 4]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                duration = video_frame_durations.pop(out_packet.pts, 0)
+                if out_packet.pts == last_video_pts:
+                    duration = max(duration, window_fill)
+                out_packet.duration = duration
+                output.mux(out_packet)
+            if out_audio is not None:
+                output.mux(out_audio.encode(None))
+        except BaseException:
+            if output is not None:
+                output.close()
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:777]: comment is valid, but the original code did the same. will ignore as not important

File: comfy_api/latest/_input_impl/video_types.py [hunk 5]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                        # the decoded pict_type would force x264's frame types (intra-only
+                        # sources like MJPEG/ProRes would come out all-keyframe)
+                        frame.pict_type = 0
+                        for out_packet in out_video.encode(frame):
+                            out_packet.duration = video_frame_durations.pop(out_packet.pts, 0)
+                            output.mux(out_packet)
+                        drain_audio()
+
+                elif packet.stream == audio_stream and not audio_done:
+                    for resampled in itertools.chain.from_iterable(map(resampler.resample, packet.decode())):
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:730]: This is technically true, but it matches the old transcode path: audio decode errors in `get_components_internal()` were not caught either, so a corrupt audio packet would abort there too

I would prefer not to change corrupt-audio behavior in this PR. Silently skipping audio packets can create gaps/drift or lower-quality output

File: comfy_api/latest/_input_impl/video_types.py [hunk 6]
```diff
@@ -394,8 +434,8 @@ def get_components_internal(self, container: InputContainer) -> VideoComponents:
 
         if len(audio_frames) > 0:
             audio_data = np.concatenate(audio_frames, axis=1)  # shape: (channels, total_samples)
-            if self.__duration:
-                audio_data = audio_data[..., :int(self.__duration * audio_stream.sample_rate)]
+            if duration:
+                audio_data = audio_data[..., :int(duration * audio_stream.sample_rate)]
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:438]: I tested a fixture where `audio_stream.sample_rate` is `0` at open because audio starts past the probe window. Once PyAV decodes audio frames, the stream sample rate is populated, and `get_components()` returns non-zero audio for both full-duration and trimmed windows around the late audio

This behavior also was not introduced by this PR - `origin/master` used the same final truncation with `audio_stream.sample_rate`; this PR only switches it to the normalized `duration`

File: comfy_api/latest/_input_impl/video_types.py [hunk 7]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                elif packet.stream == audio_stream and not audio_done:
+                    for resampled in itertools.chain.from_iterable(map(resampler.resample, packet.decode())):
+                        if not audio_started:
+                            if resampled.pts is None:
+                                frame_start = 0.0
+                            else:
+                                # passthrough frames keep the source stream's time base
+                                tb = resampled.time_base if resampled.time_base else audio_time_base
+                                frame_start = float(resampled.pts * tb)
+                            to_skip = max(0, int((start_time - frame_start) * sample_rate))
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:742]: I agree this can happen if decoded audio frames have no PTS, but I don’t think this is a practical regression from `origin/master`
I forced audio frames to `pts=None` with a temp feed proxy: the old transcode path crashes on `frame.time > ...`, while this PR drops audio for that malformed case

Handling pts-less audio correctly would require maintaining an audio timeline across resampled frames - I prefer to not add this ultra complexity for now

[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:742]: Master does the same thing :)

The old path decoded audio into a waveform and `VideoFromComponents.save_to()` writes that waveform starting at **output time 0**, so the leading gap was discarded before this PR too.

Fixing this in this PR will be a broad change and could introduce regressions and behavior change.

File: comfy_api/latest/_input_impl/video_types.py [hunk 8]
```diff
@@ -480,6 +506,254 @@ def save_to(
...
+                        logging.info("pyav decode error")
+                        continue
+                    for frame in frames:
+                        if frame.pts is not None and frame.pts < start_pts:
+                            continue
+                        if end_pts is not None and frame.pts is not None and frame.pts >= end_pts:
+                            video_done = True
+                            if last_video_pts is not None:
+                                # the source continues past the window: hold the last kept frame to the window end
+                                last_video_end = max(last_video_end, end_pts - video_pts_offset)
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py]: fixed 👍

File: comfy_api/latest/_input_impl/video_types.py [hunk 9]
```diff
@@ -58,6 +60,57 @@ def video_stream_bit_depth(stream) -> int:
...
+
+
+def probe_audio_params(container: InputContainer, audio_stream, max_packets: int = 200):
+    """Containers probed only up to a window (mpegts) leave audio codec parameters unset when
+    audio starts beyond it; learn them by decoding ahead. The caller must seek back afterwards.
+    Returns (sample_rate, channels), zeros when the stream never yields a decodable frame."""
+    for i, packet in enumerate(container.demux(audio_stream)):
+        try:
+            frames = packet.decode()
+        except av.error.FFmpegError:
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:82]: fixed 👍

File: comfy_api/latest/_input_impl/video_types.py [hunk 10]
```diff
@@ -58,6 +60,57 @@ def video_stream_bit_depth(stream) -> int:
...
+    if format != VideoContainer.AUTO and format != VideoContainer.MP4:
+        raise ValueError("Only MP4 format is supported for now")
+    if codec != VideoCodec.AUTO and codec != VideoCodec.H264:
+        raise ValueError("Only H264 codec is supported for now")
+    open_kwargs = {"mode": "w", "options": {"movflags": "use_metadata_tags"}}
+    if isinstance(format, VideoContainer) and format != VideoContainer.AUTO:
+        open_kwargs["format"] = format.value
+    elif isinstance(path, io.BytesIO):
+        open_kwargs["format"] = "mp4"  # no file extension to infer the format from
+    return open_kwargs
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:111]: it is pre-existing master behavior

File: comfy_api/latest/_input_impl/video_types.py [hunk 11]
```diff
@@ -192,13 +245,10 @@ def get_frame_count(self) -> int:
...
             # 3. Last resort: decode frames and count them (streaming)
-            if self.__start_time < 0:
-                start_time = max(self._get_raw_duration() + self.__start_time, 0)
-            else:
-                start_time = self.__start_time
+            start_time, duration = self.get_active_trim_window()
             frame_count = 1
             start_pts = int(start_time / video_stream.time_base)
-            end_pts = int((start_time + self.__duration) / video_stream.time_base)
+            end_pts = int((start_time + duration) / video_stream.time_base)
```
[the remark this answers is not in the document]
Comment (bigcat88) [reply comfy_api/latest/_input_impl/video_types.py:251]: this is the pre-existing behavior - do not want to change it in this PR

**Review by alexisrolland — APPROVED** [review]
